Override the authentication listener to validate FORM_SUBMIT

aschempp opened the pull request:
@richardhj can you see if this fixes https://github.com/contao/contao/issues/558 ?

aschempp committed:
Override the authentication listener to validate FORM_SUBMIT

Toflar reviewed (changes requested):
Concept looks logical to me. Needs tests once @richardhj provided feedback.

leofeyer commented on core-bundle/src/Security/Authentication/ContaoLoginAuthenticationListener.php:
```diff
@@ -0,0 +1,31 @@
+     */
+    protected function requiresAuthentication(Request $request)
+    {
+        if (!$request->isMethod('POST') || 0 !== strpos((string) $request->request->get('FORM_SUBMIT'), 'tl_login')) {
```
Why are you using `strpos()` here instead of a simple `===`?

aschempp replied:
the frontend login module uses `tl_login_%ID%` as the form submit value.

leofeyer replied:
Oh yes, I recall. 👍 

Personally, I would prefer to write the method like this:

```php
protected function requiresAuthentication(Request $request): bool
{
    if (!$request->isMethod('POST') || !$request->request->has('FORM_SUBMIT')) {
        return false;
    }

    return 0 === strpos($request->request->get('FORM_SUBMIT'), 'tl_login');
}
```

WDYT?

leofeyer replied:
Oh, and do we have to account for `tl_loginfoo` as well? The method would return `true` currently. Probably change it to `preg_match('/^tl_login(_\d+)?$/', $request->request->get('FORM_SUBMIT'))`?

aschempp replied:
There's no real case of `tl_loginfoo` anywhere, so I don't think we need that complication.

leofeyer committed:
Fix the coding style

leofeyer reviewed (changes requested):
The unit tests are still failing and there should be a test for the `contao.security.authentication_listener` service in the `ContaoCoreExtensionTest` class.

leofeyer committed:
Fix the coding style and add a unit test for the authentication listener

leofeyer commented on core-bundle/src/Security/Authentication/ContaoLoginAuthenticationListener.php:
```diff
@@ -0,0 +1,35 @@
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\Security\Http\Firewall\UsernamePasswordFormAuthenticationListener;
+
+class ContaoLoginAuthenticationListener extends UsernamePasswordFormAuthenticationListener
```
Why did you prefix the class name with `Contao`? It should be `LoginAuthenticationListener` IMHO.

aschempp replied:
Because the firewall listener/factory is called `contao_login`

aschempp committed:
Adjust entry point to display the login page

aschempp commented:
I have update the basic PR features according to my findings yesterday with @richardhj 

The current frontend login implementation works around several issues Contao core has with Symfony security. Symfony assumes there's a URL where you POST the login data, and that there is a page where the login form is. By default, Symfony will redirect a user to the `login_path`, and only check credentials on the `check_path`. For that case, we created the `_contao/login` route, and by setting `use_forward` the system didn't actually redirect but return a subrequest response of that page.

All of this is not actually how Contao works. In the Contao front end, one can place the login module on any page, and on that page the credentials should be checked. We also do not know a login redirect URL, we merely render the 401/403 page type *in place*.

This is what this PR finally implements now: instead of using the default authentication entry point, which redirects or internally forwards to a `login_path`, we simply generate the 401 content in that place. Instead of only checking the authentication on the `check_path`, we override the authentication listener to check whenever a POST request with FORM_SUBMIT=tl_login is sent. We therefore do not need the `contao_frontend_login` route anymore 🎉 

This PR will collide with #1164 and I will rebase and fix issues and tests once #1164 is merged.

richardhj reviewed (dismissed):
Very good! This fixes #558 and #966. Sadly #967 persists but I will evaluate once #1164 and this PR is merged.

aschempp committed:
Merge branch 'master' into feature/login-form

# Conflicts:
#	core-bundle/src/Resources/config/services.yml
#	core-bundle/src/Security/Authentication/AuthenticationEntryPoint.php

aschempp committed:
Adjust contao login security

aschempp commented:
This PR is feature-complete now, but I need to fix tests etc. Basically, we're moving away from configuring Symfony from behaving as needed by Contao, and simply implement the necessary 20% ourselves.

To summarize the changes and make them more "readable":

1. I have added a custom authentication listener (configured through the security factory). The listener will check for `$_POST['FORM_SUBMIT'] === 'tl_login'` instead of a check path. This makes `check_path` obsolete.

2. As mentioned before, our frontend login route (`/_contao/login`)  has become obsolete thanks to generating the 401 page directly in our security entry point. This makes `login_path` and `use_forward` as well as the `contao_frontend_login` route obsolete.

3. Thanks to no.2 we do no longer need a `failure_path`, because that's the place the login form is on. That's now always the current URL, and I adjusted and simplified our `AuthenticationFailureHandler` accordingly.

4. By handling base64 directly in the `AuthenticationSuccessHandler`, we do not need the `Base64RedirectController` anymore. As we **know** Contao always uses the `_target_path`, we don't need configurations for `default_target_path`.

5. The authentication success and failure handlers are now configured by the security factory, so the `success_handler` and `failure_handler` are no longer needed.

This will collide with #1130 so @bytehead needs to rebase (mostly because I already removed the lock_period etc. config from the security factory.

Toflar reviewed (commented):
Oh wow, this makes it so much easier for me to understand what's actually going on in the security authentication process 😄 Looks correct to me!

leofeyer commented on core-bundle/src/DependencyInjection/Security/ContaoLoginFactory.php:
```diff
@@ -12,40 +12,30 @@
-
-        $this->addOption('lock_period', 300);
-        $this->addOption('login_attempts', 3);
+        $this->options = ['require_previous_session' => false];
```
This unsets the other two options `check_path` and `use_forward`. Is this intended?

aschempp replied:
Yes, they will be gone in #1130 anyway. We also unset any other options, as they do not make sense for anything in Contao.

leofeyer commented on core-bundle/src/Resources/config/services.yml:
```diff
@@ -578,10 +572,23 @@ services:
         class: Contao\CoreBundle\Security\Authentication\AuthenticationFailureHandler
         arguments:
-            - '@http_kernel'
+            - '@?logger'
```
This is marked as optional dependency but it is a mandatory constructor argument in the class.

aschempp replied:
fixed in a507cf7a1e81c189e37294aa3346b3b2f0b7f77f

leofeyer committed:
Fix the coding style

leofeyer reviewed (changes requested):
PR looks very good. 👍 The tests are failing though.

aschempp committed:
Added remaining unit tests

aschempp committed:
Added some unit tests

aschempp committed:
CS

Toflar reviewed (dismissed):
LGTM

leofeyer committed:
Fix the coding style

leofeyer commented:
Thank you @aschempp.
